### PR TITLE
TGP-1600: NIPHL on cat assessments

### PR DESCRIPTION
**Pull request by yohan-oc**

### CODE REVIEW

File: app/models/ott/CategorisationInfo.scala [hunk 1]
```diff
@@ -89,22 +91,28 @@ object CategorisationInfo {
...
-        val questionsToAnswer =
+        val questionsToAnswer = {
           if (!areAllCategory1Answerable) {
             Seq.empty
           } else if (!areAllCategory2Answerable) {
             category1ToAnswer
           } else {
-            category1ToAnswer ++ category2ToAnswer
+            if (traderProfile.niphlNumber.isDefined) {
+              category1ToAnswer
```
Comment (ChrisCookOC) [inline app/models/ott/CategorisationInfo.scala]: This logic needs tweaking. Currently if you have Niphls you are never asked **any** category 2 questions, no matter if the commodity is Niphls or not


Probably needs something like below?


if (isNiphlsCommodity) 
   if (traderProfile.niphlNumber.isDefined)
      category1ToAnswer
      else 
      Seq.empty
   else
   category1ToAnswer + category2ToAnswer

File: app/navigation/Navigator.scala [hunk 2]
```diff
@@ -245,6 +245,19 @@ class Navigator @Inject() (categorisationService: CategorisationService) {
 
   private def navigateFromCategorisationPreparationPage(answers: UserAnswers, recordId: String): Call =
     answers.get(CategorisationDetailsQuery(recordId)) match {
+      case Some(catInfo)
+          if !catInfo.isNiphlAuthorised && catInfo.categoryAssessments.exists(ass => ass.isNiphlsAnswer) =>
```
Comment (ChrisCookOC) [inline app/navigation/Navigator.scala:249]: would be nice to add to the categoryInfo a function that is like
def isNiphlsAssessment = categoryAssessments.exists(ass => ass.isNiphlsAnswer) && categoryAssessments.exists(ass => ass.isCategory2 && ass.hasNoAnswers), and call that here

File: app/models/ott/CategorisationInfo.scala (hunk 1 above)
Comment (ChrisCookOC) [reply app/models/ott/CategorisationInfo.scala]: does this work?
```
        val category1Assessments = assessmentsSorted.filter(ass => ass.isCategory1)
        val category2Assessments = assessmentsSorted.filter(ass => ass.isCategory2)

        val category1ToAnswer = category1Assessments.filter(ass => !ass.hasNoAnswers).filter(ass => !ass.isNiphlsAnswer)
        val category2ToAnswer = category2Assessments.filter(ass => !ass.hasNoAnswers)

        val areAllCategory1Answerable = category1ToAnswer.size == category1Assessments.size
        val areAllCategory2Answerable = category2ToAnswer.size == category2Assessments.size

        val isNiphlsAssessment = category1Assessments.exists(ass => ass.isNiphlsAnswer) && category2Assessments.exists(ass => ass.hasNoAnswers)

        val questionsToAnswer = {
          if (isNiphlsAssessment && traderProfile.niphlNumber.isDefined) {
            category1ToAnswer
          } else if (isNiphlsAssessment) {
            Seq.empty
          } else if (!areAllCategory1Answerable) {
            Seq.empty
          } else if (!areAllCategory2Answerable) {
            category1ToAnswer
          } else {
            category1ToAnswer ++ category2ToAnswer
          }
        }
```

File: app/navigation/Navigator.scala [hunk 3]
```diff
@@ -245,6 +245,19 @@ class Navigator @Inject() (categorisationService: CategorisationService) {
 
   private def navigateFromCategorisationPreparationPage(answers: UserAnswers, recordId: String): Call =
     answers.get(CategorisationDetailsQuery(recordId)) match {
+      case Some(catInfo)
```
Comment (ChrisCookOC) [inline app/navigation/Navigator.scala]: If you make the changes suggested to category info, then none of this new logic should be needed I think

Basically the navigator doesn't really need to know whether its niphls, if we set up the "categoryAssessmentsThatNeedAnswers" properly :)

File: app/models/ott/CategorisationInfo.scala (hunk 1 above)
Comment (yohan-oc) [reply app/models/ott/CategorisationInfo.scala]: I was thinking to replacing `if (traderProfile.niphlNumber.isDefined)` with `if (isNiphlsAssessment && traderProfile.niphlNumber.isDefined)`

We will play around your above logic as well.

File: app/navigation/Navigator.scala (hunk 3 above)
Comment (yohan-oc) [reply app/navigation/Navigator.scala]: That make sense

File: app/services/CategorisationService.scala [hunk 4]
```diff
@@ -39,28 +43,68 @@ class CategorisationService @Inject() (
...
+    val hasNiphlAssessments = category1Assessments.exists(ass => ass.isNiphlsAnswer)
+
+    val category1AssessmentsWithoutNiphl = category1Assessments.filter(ass => !ass.isNiphlsAnswer)
+
+    val listOfAnswers = categorisationInfo.getAnswersForQuestions(userAnswers, recordId)
+
+    val areThereCategory1QuestionsWithNoExemption =
+      listOfAnswers.exists(x => x.answer.contains(AssessmentAnswer.NoExemption) && x.question.isCategory1)
+
+    if (categorisationInfo.isNiphlAuthorised) {
```
Comment (ChrisCookOC) [inline app/services/CategorisationService.scala]: I think the logic is slightly topsy-turvy - you only care about isNiphlAuthorised if its a niphls assessment

I think this is the same?

```
if (isNiphlsAssessment) {
      if (!categorisationInfo.isNiphlAuthorised || areThereCategory1QuestionsWithNoExemption) {
          Category1Scenario
        } else {
          Category2Scenario
        }
    } else {
      calculateResultWithoutNiphl(categorisationInfo, userAnswers, recordId)
}
```
    
    A test fails but the test data is wrong

File: test/models/ott/CategorisationInfoSpec.scala [hunk 5]
```diff
@@ -536,10 +538,181 @@ class CategorisationInfoSpec extends SpecBase {
           longerCode = true
         )
 
-        val result = CategorisationInfo.build(ottResponse, "1234567890", longerCode = true)
+        val result =
+          CategorisationInfo.build(ottResponse, "1234567890", testTraderProfileResponseWithoutNiphl, longerCode = true)
         result.value mustEqual expectedResult
       }
 
+      ".NIPHL is authorised" - {
```
Comment (ChrisCookOC) [inline test/models/ott/CategorisationInfoSpec.scala]: Should perhaps be a section for testing Niphls cases whether or not user is authorised for Niphls

There should be a different answer if they don't have niphls but it is a niphls assessment (when with or without extra assessments)

File: test/services/CategorisationServiceSpec.scala [hunk 6]
```diff
@@ -307,6 +313,110 @@ class CategorisationServiceSpec extends SpecBase with BeforeAndAfterEach {
 
       }
 
+      "if NIPHL is not authorised and has NIPHL assessments" in {
```
Comment (ChrisCookOC) [inline test/services/CategorisationServiceSpec.scala:332]: this is not a valid NIPHLs assessment

File: test/services/CategorisationServiceSpec.scala [hunk 7]
```diff
@@ -307,6 +313,110 @@ class CategorisationServiceSpec extends SpecBase with BeforeAndAfterEach {
...
+          )
+        )
+
+        val categorisationInfo = CategorisationInfo(
+          "1234567890",
+          Seq(
+            assessment1,
+            assessment2
+          ),
+          Seq(assessment1, assessment2),
```
Comment (ChrisCookOC) [inline test/services/CategorisationServiceSpec.scala]: This should be empty as both items are Niphls and so do not need answering

File: test/services/CategorisationServiceSpec.scala [hunk 8]
```diff
@@ -307,6 +313,110 @@ class CategorisationServiceSpec extends SpecBase with BeforeAndAfterEach {
...
+        )
+
+        val userAnswers = emptyUserAnswers
+          .set(CategorisationDetailsQuery(testRecordId), categorisationInfo)
+          .success
+          .value
+          .set(AssessmentPage(testRecordId, 0), AssessmentAnswer.Exemption)
+          .success
+          .value
+          .set(AssessmentPage(testRecordId, 1), AssessmentAnswer.NoExemption)
```
Comment (ChrisCookOC) [inline test/services/CategorisationServiceSpec.scala]: These should not have been answered

File: test/services/CategorisationServiceSpec.scala [hunk 9]
```diff
@@ -307,6 +313,110 @@ class CategorisationServiceSpec extends SpecBase with BeforeAndAfterEach {
...
+        )
+
+        val categorisationInfo = CategorisationInfo(
+          "1234567890",
+          Seq(
+            assessment1,
+            assessment2,
+            assessment3
+          ),
+          Seq(assessment1, assessment2, assessment3),
```
Comment (ChrisCookOC) [inline test/services/CategorisationServiceSpec.scala]: Not valid Niphls assessment

File: app/navigation/Navigator.scala (hunk 3 above)
Comment (yohan-oc) [reply app/navigation/Navigator.scala]: This now removed. But we kept the unit tests related NIPHLs

File: app/models/ott/CategorisationInfo.scala (hunk 1 above)
Comment (yohan-oc) [reply app/models/ott/CategorisationInfo.scala]: Updated to suggested logic and fixed unit tests

File: test/services/CategorisationServiceSpec.scala (hunk 7 above)
Comment (yohan-oc) [reply test/services/CategorisationServiceSpec.scala]: Fixed

File: test/services/CategorisationServiceSpec.scala (hunk 8 above)
Comment (yohan-oc) [reply test/services/CategorisationServiceSpec.scala]: Fixed

File: test/services/CategorisationServiceSpec.scala (hunk 9 above)
Comment (yohan-oc) [reply test/services/CategorisationServiceSpec.scala]: corrected

File: test/services/CategorisationServiceSpec.scala (hunk 6 above)
Comment (yohan-oc) [reply test/services/CategorisationServiceSpec.scala:332]: This is for the code coverage. Now removed not needed with new logic.

File: test/models/ott/CategorisationInfoSpec.scala (hunk 5 above)
Comment (yohan-oc) [reply test/models/ott/CategorisationInfoSpec.scala]: Tests updated

File: app/models/ott/CategorisationInfo.scala [hunk 10]
```diff
@@ -29,7 +29,8 @@ final case class CategorisationInfo(
   categoryAssessmentsThatNeedAnswers: Seq[CategoryAssessment],
   measurementUnit: Option[String],
   descendantCount: Int,
-  longerCode: Boolean = false
+  longerCode: Boolean = false,
+  isNiphlAuthorised: Boolean = false
```
Comment (ChrisCookOC) [inline app/models/ott/CategorisationInfo.scala]: can it be isTraderNiphlsAuthorised. Just make it clear its the trader who we are talking about here

File: test/models/ott/CategorisationInfoSpec.scala [hunk 11]
```diff
@@ -536,10 +538,277 @@ class CategorisationInfoSpec extends SpecBase {
...
+                None,
+                0,
+                isNiphlAuthorised = true
+              )
+
+            val result = CategorisationInfo.build(mockOttResponse, "1234567890", testTraderProfileResponseWithNiphl)
+            result.value mustEqual expectedResult
+          }
+
+          "when there are Category 1 that needed answer and there is a NIPHL assessment but there is a Category 2 assessment with no exemptions" in {
```
Comment (ChrisCookOC) [inline test/models/ott/CategorisationInfoSpec.scala]: pedantic: but should be and in the title

File: test/navigation/NavigatorSpec.scala [hunk 12]
```diff
@@ -589,19 +589,45 @@ class NavigatorSpec extends SpecBase with BeforeAndAfterEach {
...
 
         "must go from categorisation preparation" - {
 
-          "to category guidance page when assessments need answering" in {
-            val userAnswers = emptyUserAnswers
-              .set(CategorisationDetailsQuery(testRecordId), categorisationInfo)
-              .success
-              .value
+          "to category guidance page" - {
+            "if NIPHL is not  authorised and assessments need answering" in {
```
Comment (ChrisCookOC) [inline test/navigation/NavigatorSpec.scala]: pedantic: no need to mention niphls in tests that aren't testing niphls

File: test/navigation/NavigatorSpec.scala [hunk 13]
```diff
@@ -660,6 +686,60 @@ class NavigatorSpec extends SpecBase with BeforeAndAfterEach {
 
             }
 
+            "NIPHL is not authorised and has NIPHL assesments" in {
+
+              val categoryInfoWithNiphlAssessments = CategorisationInfo(
+                "1234567890",
+                Seq(category1Niphl),
```
Comment (ChrisCookOC) [inline test/navigation/NavigatorSpec.scala]: Missing the empty cat 2 needed for a NIPHLs assessment

File: test/navigation/NavigatorSpec.scala [hunk 14]
```diff
@@ -660,6 +686,60 @@ class NavigatorSpec extends SpecBase with BeforeAndAfterEach {
...
+              navigator.nextPage(
+                CategorisationPreparationPage(testRecordId),
+                NormalMode,
+                userAnswers
+              ) mustBe routes.CategorisationResultController
+                .onPageLoad(testRecordId, Category1Scenario)
+
+            }
+
+            "NIPHL is authorised and has one NIPHL assessments and category 2 assessment with no exemptions" in {
```
Comment (ChrisCookOC) [inline test/navigation/NavigatorSpec.scala:716]: Possible additional case. Niphl is not authorised and has one Niphl assessment and category 2 no exempt and other category 1 questions

File: test/navigation/NavigatorSpec.scala [hunk 15]
```diff
@@ -901,6 +981,33 @@ class NavigatorSpec extends SpecBase with BeforeAndAfterEach {
 
             }
 
+            "if NIPHL is authorised and has NIPHL assessment and the answers are yes for all category 1 assessments" - {
+
+              val categoryInfoWithNiphlAssessments = CategorisationInfo(
+                "1234567890",
+                Seq(category1Niphl, category1),
+                Seq(category1Niphl, category1),
```
Comment (ChrisCookOC) [inline test/navigation/NavigatorSpec.scala]: Not valid scenario. Shouldn't ask Niphls question

File: test/navigation/NavigatorSpec.scala [hunk 16]
```diff
@@ -1260,6 +1367,30 @@ class NavigatorSpec extends SpecBase with BeforeAndAfterEach {
               routes.CategorisationResultController.onPageLoad(testRecordId, Category2Scenario)
           }
 
+          ".NIPHL assesments" - {
+
+            "to category 2 result page when categorisation result is so" in {
+              val categoryInfoWithNiphlAssessments = CategorisationInfo(
```
Comment (ChrisCookOC) [inline test/navigation/NavigatorSpec.scala]: This isn't actually testing anything different from the current Category2Scenario test, because the mock is driving the behaviour rather than the UA?

File: app/models/ott/CategorisationInfo.scala (hunk 10 above)
Comment (yohan-oc) [reply app/models/ott/CategorisationInfo.scala]: This is renamed

File: app/services/CategorisationService.scala (hunk 4 above)
Comment (yohan-oc) [reply app/services/CategorisationService.scala]: Logic updated and fix failing tests

File: test/navigation/NavigatorSpec.scala (hunk 14 above)
Comment (yohan-oc) [reply test/navigation/NavigatorSpec.scala:716]: Added a new test

File: test/navigation/NavigatorSpec.scala (hunk 16 above)
Comment (yohan-oc) [reply test/navigation/NavigatorSpec.scala]: Removed

File: test/navigation/NavigatorSpec.scala (hunk 13 above)
Comment (yohan-oc) [reply test/navigation/NavigatorSpec.scala]: Updated

File: test/navigation/NavigatorSpec.scala (hunk 12 above)
Comment (yohan-oc) [reply test/navigation/NavigatorSpec.scala]: Title updated

File: test/models/ott/CategorisationInfoSpec.scala (hunk 11 above)
Comment (yohan-oc) [reply test/models/ott/CategorisationInfoSpec.scala]: Updated

File: test/navigation/NavigatorSpec.scala (hunk 15 above)
Comment (yohan-oc) [reply test/navigation/NavigatorSpec.scala]: Updated

File: app/services/CategorisationService.scala [hunk 17]
```diff
@@ -39,28 +43,61 @@ class CategorisationService @Inject() (
...
-        Some(recordId),
-        country,
-        LocalDate.now()
-      )
-      .flatMap { response =>
-        CategorisationInfo.build(response, commodityCode, longerCode) match {
-          case Some(categorisationInfo) => Future.successful(categorisationInfo)
-          case _                        =>
-            Future.failed(new RuntimeException("Could not build categorisation info"))
+    profileConnector.getTraderProfile(request.eori).flatMap { profile =>
```
Comment (ChrisCookOC) [inline app/services/CategorisationService.scala:45]: Test case could be added for if getTraderProfile fails

Comment (yohan-oc) [reply app/services/CategorisationService.scala:45]: Added a new test

**Comment by ChrisCookOC** [comment]
Won't approve as I've just pushed fixes to some edge cases, but I'm happy :)

**Review by Ewan-James-Donovan — APPROVED** [review]
